File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        
        # O(n^2) time with O(n) space
        
        # create 19 hash maps
        cols = [{} for i in range(9)]
        
        squareMaps = {(0, 0): {}, (0, 1): {}, (0, 2): {},
                      (1, 0): {}, (1, 1): {}, (1, 2): {},
                      (2, 0): {}, (2, 1): {}, (2, 2): {}}
        
        
        for i, row in enumerate(board):
            row_hmap = {}
            for j, item in enumerate(row):
                # is item .?
                if item == ".":
                    continue
                  
                # is item in row hmap?
                if row_hmap.get(item, None):
                    return False
                # is item in corresponding col?
                elif cols[j].get(item, None):
                    return False
                
                squareMap = squareMaps[(i // 3, j // 3)]
                # is item in corresponding square?
                if squareMap.get(item, None):
                    return False
                
                row_hmap[item] = 1
                cols[j][item] = 1
                squareMap[item] = 1
                
                
        return True

File: 0036-valid-sudoku/0036-valid-sudoku.cs
public class Solution {
    public bool IsValidSudoku(char[][] board) {
        Dictionary<(int, int), HashSet<int>> quad = new Dictionary<(int, int), HashSet<int>>();
        List<HashSet<int>> row, col;
        
        row = new List<HashSet<int>>();
        col = new List<HashSet<int>>();
        
        for(int i = 0; i < 9; i++){
            row.Add(new HashSet<int>());
            col.Add(new HashSet<int>());
            
            for(int j = 0; j < 9; j ++){
                quad.Add((i, j), new HashSet<int>());
            }
        }
        
        for(int i=0; i < 9; i++){
            for(int j=0; j< 9; j++){
                char num = board[i][j];
                if(num == '.') { continue; }
                (int, int) q = ((int) Math.Floor((double)(i / 3)), (int) Math.Floor((double)(j / 3)));
                if(quad[q].Contains(num) || row[i].Contains(num) || col[j].Contains(num)){
                    return false;
                }
                quad[q].Add(num);
                row[i].Add(num);
                col[j].Add(num);
            }
        }
        
        return true;
        
    }
}